### PR TITLE
Core: Refactors to BinPack and Rewrite APIs

### PR DESCRIPTION
**Pull request by RussellSpitzer**
Changes to support the implementation of the Spark RewriteDatafileAciton. Fixes JavaDoc links,
adds in BinPackStrategy comments to better binpack real world usecases.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java [hunk 1]
```diff
@@ -137,6 +139,78 @@ public RewriteStrategy options(Map<String, String> options) {
...
+   * If we are writing 10.1 G of data with a target file size of 1G we would end up with
+   * 11 files, one of which would only have 0.1g. This would most likely be less preferable to
+   * 10 files each of which was 1.01g. So here we decide whether to round up or round down
+   * based on what the estimated average file size will be if we ignore the remainder (0.1g). If
+   * the new file size is less than 10% greater than the target file size then we will round down
+   * when determining the number of output files.
+   * @param totalSizeInBytes total data size for a file group
+   * @return the number of files this strategy should create
+   */
+  protected long numOutputFiles(long totalSizeInBytes) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:159]: This logic looks correct to me.

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java [hunk 2]
```diff
@@ -137,6 +139,78 @@ public RewriteStrategy options(Map<String, String> options) {
...
+
+    long fileCountWithRemainder = LongMath.divide(totalSizeInBytes, targetFileSize, RoundingMode.CEILING);
+    if (LongMath.mod(totalSizeInBytes, targetFileSize) > minFileSize) {
+      // Our Remainder file is of valid size for this compaction so keep it
+      return fileCountWithRemainder;
+    }
+
+    long fileCountWithoutRemainder = LongMath.divide(totalSizeInBytes, targetFileSize, RoundingMode.FLOOR);
+    long avgFileSizeWithoutRemainder = totalSizeInBytes / fileCountWithoutRemainder;
+    if (avgFileSizeWithoutRemainder < 1.1 * targetFileSize) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:172]: Am I correct that `numOutputFiles` handles the case when `1.1 * targetFileSize` is larger that the max value?

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java [hunk 3]
```diff
@@ -137,6 +139,78 @@ public RewriteStrategy options(Map<String, String> options) {
...
+   */
+  protected long splitSize(long totalSizeInBytes) {
+    long estimatedSplitSize = totalSizeInBytes / numOutputFiles(totalSizeInBytes);
+    return Math.min(estimatedSplitSize, writeMaxFileSize());
+  }
+
+  protected static long inputFileSize(List<FileScanTask> fileToRewrite) {
+    return fileToRewrite.stream().mapToLong(FileScanTask::length).sum();
+  }
+
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:193]: nit: extra line

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java [hunk 4]
```diff
@@ -137,6 +139,78 @@ public RewriteStrategy options(Map<String, String> options) {
...
+  /**
+   * Returns the smaller of our max write file threshold, and our estimated split size based on
+   * the number of output files we want to generate.
+   */
+  protected long splitSize(long totalSizeInBytes) {
+    long estimatedSplitSize = totalSizeInBytes / numOutputFiles(totalSizeInBytes);
+    return Math.min(estimatedSplitSize, writeMaxFileSize());
+  }
+
+  protected static long inputFileSize(List<FileScanTask> fileToRewrite) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:190]: This seems the only one static method. Is that on purpose?

**Review by aokolnychyi — APPROVED** [review]
LGTM. Just a few questions.

Thanks for the Javadoc around thresholds, @RussellSpitzer!

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java (hunk 4 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:190]: No reason, i'll remove it

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java (hunk 2 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:172]: it isn't here, i'm not sure how we want to settle this. We could do say that if the avgFileSizeWithoutRemainder is < maxFileSize? I think we discussed that previously

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:172]: I think we are getting into edge cases here, but maybe we could do Math.min(1.1 * targetFileSize, maxWriteSize)

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:172]:
```java
    long fileCountWithoutRemainder = LongMath.divide(totalSizeInBytes, targetFileSize, RoundingMode.FLOOR);
    long avgFileSizeWithoutRemainder = totalSizeInBytes / fileCountWithoutRemainder;
    if (avgFileSizeWithoutRemainder < Math.min(1.1 * targetFileSize, writeMaxFileSize())) {
      // Round down and distribute remainder amongst other files
      return fileCountWithoutRemainder;
    } else {
      // Keep the remainder file
      return fileCountWithRemainder;
    }
```

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:172]: `Math.min(1.1 * targetFileSize, maxWriteSize)` seems reasonable to me.

**Comment by RussellSpitzer** [comment]
https://github.com/apache/iceberg/pull/2770 - Moving to this PR because I accidentally made this branch inside the apache repo

**Pull request closed without merging**